cmd/apiserver-watcher/gcp_handler.go

```diff
@@ -0,0 +1,209 @@
+package main
+
+import (
+	"encoding/json"
+	"net"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	"cloud.google.com/go/compute/metadata"
+	"github.com/coreos/go-iptables/iptables"
+	"github.com/golang/glog"
+	"github.com/vishvananda/netlink"
+)
+
+// Update iptables rules based on google cloud load balancer VIPS
+//
+// This is needed because the GCP L3 load balancer doesn't actually do DNAT;
+// the destination IP address is still the VIP. Normally, there is an agent that
+// adds the vip to the local routing table, tricking the kernel in to thinking
+// it's a local IP and allowing processes doing an accept(0.0.0.0) to receive
+// the packets. Clever.
+//
+// We don't do that. Instead, we DNAT with conntrack. This is so we don't break
+// existing connections when the vip is removed. This is useful for draining
+// connections - take ourselves out of the vip, but service existing conns.
+
+const (
+	// FIXME: support multiple interfaces
+	gcpMetadataForwardedIPs = "instance/network-interfaces/0/forwarded-ips/?recursive=true"
+	// name of nat chain for iptables masquerade rules
+	gcpMasqChainName      = "gcp-vips"
+	gcpLocalMasqChainName = "gcp-vips-local"
+)
+
+type gcpHandler struct {
+	vips      []string
+	ip4tables *iptables.IPTables
+	// GCP doesn't support IPv6
+	// ip6tables *iptables.IPTables
+	stopCh chan struct{}
+}
+
+var _ handler = &gcpHandler{}
+
+func newGCPHandler() (handler, error) {
+	ipt4, err := iptables.New(
+		iptables.IPFamily(iptables.ProtocolIPv4),
+		iptables.Timeout(5),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return &gcpHandler{
+		ip4tables: ipt4,
+		stopCh:    make(chan struct{}),
+	}, nil
+}
+
+// onFailure: either stop the routes service, or write downfile
+func (gcp *gcpHandler) onFailure() error {
+	select {
+	case gcp.stopCh <- struct{}{}:
+		glog.V(4).Info("stopping iptables sync loop")
+	default:
+		glog.V(4).Info("no sync iptables running")
+	}
+	// wait for iptables sync loop to finish
+	time.Sleep(iptablesTimeout * time.Second)
+	return gcp.deleteRules()
+}
+
+// onSuccess: either start routes service, or remove down file
+func (gcp *gcpHandler) onSuccess() error {
+	if err := gcp.syncRulesOnce(); err != nil {
+		return err
+	}
+
+	// bz1930457 delete stale conntrack entries to the apiserver
+	// FIXME: assume IPv4 only because GCP doesn't support IPv6
+	for _, vip := range gcp.vips {
+		// sanity check to not flush unexpected IPs
+		// provided by the metadata service
+		ip := net.ParseIP(vip)
+		if ip == nil || !ip.IsGlobalUnicast() {
+			continue
+		}
+		filter := &netlink.ConntrackFilter{}
+		filter.AddIP(netlink.ConntrackNatSrcIP, ip)
```

Comment:
This is deprecated: https://github.com/vishvananda/netlink/blob/master/conntrack_linux.go#L337

We definitely want ConntrackReplySrcIP here.

Also, we need to ensure the conntrack steps are applied after iptables rules sync, not before, so we don't race against ourselves.

Reply:
> This is deprecated: https://github.com/vishvananda/netlink/blob/master/conntrack_linux.go#L337
> We definitely want ConntrackReplySrcIP here.

We are vendoring the netlink stable release v1.0.0, and the code is backwards compatible, ConntrackReplySrcIP is an alias, we can bump the library if we want to have the new name and bring the modifications I've added to the conntrack filtering https://github.com/vishvananda/netlink/commit/bca67dfc8220b44ef582c9da4e9172bf1c9ec973 , but do we really need this or is  a nice to have?

> Also, we need to ensure the conntrack steps are applied after iptables rules sync, not before, so we don't race against ourselves.

yeah, it is how is implemented, check the onSuccess() methods:
syncOnce() do a first sync, if errors, it return,s otherwise it flush the conntrack entries (iptables rules are there already) and then it syncs rules in a loop until the stop signal is sent onFailure()

Reply:
As long as we call the conntrack bits whenever we add the rules, we are good.

We should use whatever version of netlink other parts of OpenShift are using for the time being, I guess.  1.1.0 is the latest tag I see there, but even that branch is over a year old now.  I'll leave that decision up to you.